### DISCUSSION TITLE
Pessimistic locking in hotrod remote cache api?

### DISCUSSION THREAD

**Discussion by aaronriekenberg** [infinispan/infinispan#13035] [Q&A]
In Hazelcast it is possible to do [pessimistic locking on cache map by key](https://docs.hazelcast.com/hazelcast/5.5/data-structures/locking-maps). The call to lock() on the map blocks until the lock is acquired.  It is possible to set a timeout for this call.

I am trying to do something similar with Infinispan RemoteCache using HotRod Java APIs.

Here is what i have tried so far:
1. Using transactional cache configuration configured to PESSIMISTIC locking as [shown here](https://infinispan.org/docs/stable/titles/hotrod_java/hotrod_java.html#hr_transactions_config_server).  With this design, I observe if multiple threads try to update the same cache key concurrently, I get RollbackExceptions when calling `TransactionManager.commit()` and have to retry many times.
2. Using counters as a lock - I use `StrongCounter.sync().compareAndSet(0, 1)` to acquire the lock and `StrongCounter.sync().reset()` to release the lock.   This works and does not throw exceptions, but `compareAndSet` may return false so have to keep looping many times to retry.

Is there any Remote Cache API similar to [Hazelcast IMap.lock(key)](https://docs.hazelcast.org/docs/5.5.1/javadoc/com/hazelcast/map/IMap.html#lock(K)) which blocks until the lock is acquired and does not need looping?

Thanks!

**Comment by pruivo** [comment]
Short answer, no.

This blog post is not mine but I share the same opinion: https://belaban.blogspot.com/2020/11/i-hate-distributed-locks.html

There are tons of problems with distributed locking making them useless. It is fine if you are in a perfect world, but not on this one. When things fail, you have a decision to make:
* allow locks to expire if acquired for too long, which means you can have one or more applications accessing the resources. If that case is valid, why use locks!?
* block everything indefinitely (I think Hazelcast does this). Performance will be impacted if, for example, an application keeps the locks for 2 days!

We don't want to make the choice, so it is up to you to decide and face the consequences. 

If you need to change keys in multiple (or just a single) Infinispan cache, transactions are the way to go. Locks are acquired and commit time and the transaction is replayed on the server side. As with any transactional system, do not have side effects on your transactions :)
